File: Services/dotnet-services/Core/OmniCure.Core.Domain/Interfaces/Persistence/Repositories/IWriteRepository.cs
using System.Linq.Expressions;

namespace OmniCure.Core.Domain.Interfaces.Persistence.Repositories;

public interface IWriteRepository<T> where T : class
{
    Task<T?> GetByIdAsync(Guid id, CancellationToken cancellationToken = default);
    
    Task<T?> GetByIdAsync<TKey>(TKey id, CancellationToken cancellationToken = default);
    
    Task<T> GetFirstAsync(Expression<Func<T, bool>> predicate, CancellationToken cancellationToken = default);

    Task<T?> GetFirstOrDefaultAsync(Expression<Func<T, bool>> predicate, CancellationToken cancellationToken = default);

    Task AddAsync(T entity, CancellationToken cancellationToken = default);

    void Update(T entity);

    void Remove(T entity, bool hardDelete = false);

    Task AddRangeAsync(IEnumerable<T> entities, CancellationToken cancellationToken = default);

    Task<List<T>> GetListAsync(Expression<Func<T, bool>> predicate, CancellationToken cancellationToken = default);

    Task<List<T>> GetAsync(
        Expression<Func<T, bool>>? filter = null,
        string? sortBy = null,
        string? sortDirection = "asc",
        CancellationToken cancellationToken = default);
    
    IQueryable<T> GetQueryable();
    
    void RemoveRange(IEnumerable<T> entities, bool hardDelete = false);

    Task<bool> ExistsAsync(Expression<Func<T, bool>> predicate, CancellationToken cancellationToken = default);
}

File: Services/java-services/Core/CoreDomain/src/main/java/ai/omnicure/core/domain/repository/IWriteRepository.java
package ai.omnicure.core.domain.repository;

import java.util.List;
import java.util.Optional;

public interface IWriteRepository<T, ID> {
    
    Optional<T> getById(ID id);
    
    T add(T entity);
    
    T update(T entity);
    
    void remove(T entity, boolean hardDelete);
    
    List<T> addRange(Iterable<T> entities);
    
    void removeRange(Iterable<T> entities, boolean hardDelete);

    boolean existsById(ID id);
}
